### ISSUE TITLE
Aggregation methods, polygon tables

### ISSUE THREAD

**Issue by njmattes** [RDCEP/EDE#22]
@ricardobarroslourenco Once we have a finalized procedure for returning a raster of values, we should finish at least one procedure for aggregating those to polygons. But we don't yet have polygons, I just realized. We need to sort out a way to store these ASAP. @legendOfZelda Do you know if plenario has an ingestor for Shapefiles already in place?

**Comment by ricardobarroslourenco** [comment]
@njmattes there is a `shp2pgsql` (such as the `raster2pgsql`) described in a [tutorial](http://revenant.ca/www/postgis/workshop/loadingdata.html). I saw that is also possible to get a `GeoJSON` compliant polygon and ingest. The shapefiles to ingest would be the `GADM` regions?

**Comment by njmattes** [comment]
Yes I think we can start with just the GADM0 regions.Unless it's easier to ingest all of them (GADM1, GADM2) at once. Joshua's mentioned FPU regions, but I don't have outlines for those.

**Comment by ricardobarroslourenco** [comment]
@njmattes we could load them in batches. Specially if something goes wrong, and we need to rollback. Do we have all those shapefiles already downloaded, or we still need to do so at the [GADM website](http://www.gadm.org/country)?

**Comment by njmattes** [comment]
I don't have the GADM files anymore (they're huge). You can download the entire world here: [http://www.gadm.org/version2](http://www.gadm.org/version2).

**Comment by ricardobarroslourenco** [comment]
@njmattes  I've created a folder at `/var/www/gadm` and downloaded the entire world there (by the way, super fast - 334MB in 14s). It is still zipped, and I would like to hear from @legendOfZelda if he tried to ingest shapefiles on his tests, and if we need to change something prior to ingest.

**Comment by legendOfZelda** [comment]
under construction...

**Comment by legendOfZelda** [comment]
i ingested the `.shp` file into a table called `regions` using `ogr2ogr`. the table with its columns was created by `ogr2ogr` itself but we don't really want all these columns. all we need is a primary key, geom, and a JSON with the metadata (that can contain the geometry again, not super efficient in space but is convenient).

**Comment by ricardobarroslourenco** [comment]
As said on monday, for this prototype, it is ok to use `ogr2ogr`. But on a further iteration should be interesting to replace into a more customized function, to avoid the generation of blank columns.

**Comment by legendOfZelda** [comment]
how about this schema here?

```
CREATE TABLE regions_meta (
    uid         bigserial primary key,
    name        text,   # e.g. GADM, EEZ
    version     text,   # for e.g. GADM: 1, 2.0, 2.7, 2.8 ;
                        # for e.g. EEZ: 1, 2, 3, ..., 6, 6.1, 7, 8
    attributes  text[]  # for e.g. GADM v2.8:   ID_0, NAME_0, VARNAME_0, TYPE_0, ...
                        # for e.g. EEZ  v8:     ID, OBJECTID_1, EEZ, Country, Sovereign, Remarks, ...
);

CREATE TABLE regions (
    uid         bigserial primary key,
    geom        geometry,
    meta_data   jsonb,  # key-value pairs with keys = attributes from regions_meta
    meta_id     bigint references regions_meta(uid)
);
```

to support `GADM`, `EEZ` (Exclusive Economic Zones Boundaries), and others, e.g. the ones mentioned in http://www.gadm.org/links

**Comment by legendOfZelda** [comment]
i am almost done with writing the ingestion script for shapefiles such as for GADM, EEZ (exclusive economic zones), etc. once we have these tables filled up we can specify polygons by ID which again is convenient when specifying the flask URL.

**Comment by legendOfZelda** [comment]
ok, ingestion into above 2 tables works, just ingested `EEZ` (`GADM` works too but it just takes too long as for now). now i can continue with the flask routes and within the urls i can conveniently use polygon IDs.

**Comment by ricardobarroslourenco** [comment]
Nice. How long did it take to load? I just arrived at the CI and I'll be reviewing stuff.

**Comment by legendOfZelda** [comment]
took roughly 5' to ingest `EEZ` but im still ingesting one polygon at a time, so lots of potential to speed it up. it's all yet still in the `severin` branch, not yet merged into `develop`. the new tables are, as usual, in `models.py` and the shapefile ingestion script is `ingest/ingest_shapes.py`.

**Comment by ricardobarroslourenco** [comment]
Ok. I'll look into the `severin` branch. I'll be also reviewing the architecture we are using as @njmattes asked on monday.

**Comment by ricardobarroslourenco** [comment]
@legendOfZelda about the `/ede/cache_builder.py`, is it a cache that gets records in space, but caching all time frames that are present?

**Comment by legendOfZelda** [comment]
hm, haven't looked into `cache_builder.py` yet, but we can discuss it today.
